internal/controller/gateway.go

```diff
@@ -31,46 +31,103 @@ import (
-// to watch for Gateway objects across all namespaces.
-func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger) (controller.Controller, error) {
+// to watch for Gateway objects across all namespaces and reconcile those that match class.
+func NewGatewayController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger, gatewayClassControllerName string) (controller.Controller, error) {
 	r := &gatewayReconciler{
-		client:       mgr.GetClient(),
-		eventHandler: eventHandler,
-		FieldLogger:  log,
+		ctx:                        context.Background(),
+		client:                     mgr.GetClient(),
+		eventHandler:               eventHandler,
+		log:                        log,
+		gatewayClassControllerName: gatewayClassControllerName,
 	}
 	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return nil, err
 	}
-	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.Gateway{}}, r.enqueueRequestForOwnedGateway()); err != nil {
 		return nil, err
 	}
 	return c, nil
 }
 
+// enqueueRequestForOwnedGateway returns an event handler that maps events to
+// Gateway objects that reference a GatewayClass owned by Contour.
+func (r *gatewayReconciler) enqueueRequestForOwnedGateway() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		gw, ok := a.(*gatewayapi_v1alpha1.Gateway)
+		if !ok {
+			r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+				Info("invalid object, bypassing reconciliation.")
+			return []reconcile.Request{}
+		}
+		if matches, err := r.hasContourOwnedClass(gw); matches {
+			if err != nil {
+				r.log.WithField("namespace", gw.Namespace).WithField("name", gw.Name).Error(err)
+				return []reconcile.Request{}
+			}
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: gw.Namespace,
+						Name:      gw.Name,
+					},
+				},
+			}
+		}
+		r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+			Info("configured controllerName doesn't match an existing GatewayClass")
+		return []reconcile.Request{}
+	})
+}
+
+// hasContourOwnedClass returns true if the class referenced by gateway exists
+// and matches the configured controllerName.
+func (r *gatewayReconciler) hasContourOwnedClass(gw *gatewayapi_v1alpha1.Gateway) (bool, error) {
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(r.ctx, types.NamespacedName{Name: gw.Spec.GatewayClassName}, gc); err != nil {
+		return false, fmt.Errorf("failed to get gatewayclass %s: %w", gw.Spec.GatewayClassName, err)
+	}
+	if gc.Spec.Controller != r.gatewayClassControllerName {
+		return false, nil
+	}
+	return true, nil
```

Comment:
We should check here that the GatewayClass is also Admitted, although this adds a race condition with the status updater. We'll probably need a boolean stored inside the currently-running Contour somewhere that indicates if it's successfully admitted a gatewayclass, and maybe its name as well?

This could conceivably be a followup item though, you'd have to work pretty hard to do this, and it would produce weird outcomes. If we *don't* check for `Admitted` on the GatewayClass in this iteration, then we should document that as a limitation of the current implementation, and say that you shouldn't do that in the release notes.

Reply:
The above is more complicated than it might seem because we have to handle the case that the GatewayClass changes once Contour's admitted it, and drops out of being the admitted GatewayClass.

That's another case that should be covered by tests, if it's not already (still reviewing), what's the behavior when a GatewayClass becomes unadmitted but there are valid Gateways associated with it?

Reply:
Let's add a follow-up. I can see if we add this check then we'll need a way to resync the Gateway. For instance, if a Gateway is created, and the Class is not yet Admitted, we'll skip and wait for that status. But when the Class is finally admitted, then we should resync for Gateways to check for any that are valid and Admit those, but there won't be an object update so we'll need to query the API for Gateways.